File: backend/src/services/ledgerService.js
import { config } from "../config.js";
import { readJson, writeJson } from "../utils/fileStore.js";
import { buildLedgerHash } from "../utils/hash.js";

class LedgerService {
  getEntries() {
    return readJson(config.ledgerPath, []);
  }

  saveEntries(entries) {
    writeJson(config.ledgerPath, entries);
  }

  validateChain() {
    const entries = this.getEntries();

    for (let index = 0; index < entries.length; index += 1) {
      const entry = entries[index];
      const expectedPreviousHash = index === 0 ? "GENESIS" : entries[index - 1].CurrentHash;
      const expectedCurrentHash = buildLedgerHash({
        ...entry,
        PreviousHash: expectedPreviousHash
      });

      if (entry.PreviousHash !== expectedPreviousHash || entry.CurrentHash !== expectedCurrentHash) {
        return {
          valid: false,
          index,
          expectedPreviousHash,
          expectedCurrentHash,
          actualPreviousHash: entry.PreviousHash,
          actualCurrentHash: entry.CurrentHash
        };
      }
    }

    return { valid: true, index: -1 };
  }

  appendEntry(entryWithoutHashes) {
    const entries = this.getEntries();
    const previousHash = entries.length === 0 ? "GENESIS" : entries[entries.length - 1].CurrentHash;
    const nextEntry = {
      ...entryWithoutHashes,
      PreviousHash: previousHash
    };

    nextEntry.CurrentHash = buildLedgerHash(nextEntry);
    entries.push(nextEntry);
    this.saveEntries(entries);

    return nextEntry;
  }
}

export const ledgerService = new LedgerService();

File: backend/flask_app/services/ledger_service.py
from flask import current_app

from ..utils.file_store import read_json, write_json
from ..utils.hash_utils import build_ledger_hash


class LedgerService:
    def get_entries(self):
        return read_json(current_app.config["LEDGER_PATH"], [])

    def save_entries(self, entries):
        write_json(current_app.config["LEDGER_PATH"], entries)

    def validate_chain(self):
        entries = self.get_entries()

        for index, entry in enumerate(entries):
            expected_previous_hash = "GENESIS" if index == 0 else entries[index - 1]["CurrentHash"]
            expected_current_hash = build_ledger_hash({**entry, "PreviousHash": expected_previous_hash})

            if entry["PreviousHash"] != expected_previous_hash or entry["CurrentHash"] != expected_current_hash:
                return {
                    "valid": False,
                    "index": index,
                    "expectedPreviousHash": expected_previous_hash,
                    "expectedCurrentHash": expected_current_hash,
                    "actualPreviousHash": entry["PreviousHash"],
                    "actualCurrentHash": entry["CurrentHash"],
                }

        return {"valid": True, "index": -1}

    def append_entry(self, entry_without_hashes):
        entries = self.get_entries()
        previous_hash = "GENESIS" if not entries else entries[-1]["CurrentHash"]
        next_entry = {**entry_without_hashes, "PreviousHash": previous_hash}
        next_entry["CurrentHash"] = build_ledger_hash(next_entry)
        entries.append(next_entry)
        self.save_entries(entries)
        return next_entry


ledger_service = LedgerService()
